plat/kvm/x86/setup.c

```diff
@@ -1,456 +1,295 @@
 
-static void _init_paging(struct multiboot_info *mi)
-{
-	struct kvmplat_config_memregion *mr[2];
-	multiboot_memory_map_t *m;
-	__sz offset, len;
-	__paddr_t start, paddr;
-	__vaddr_t vaddr;
-	__sz free_memory, res_memory;
-	unsigned long frames;
-	int rc;
+		vaddr = PAGE_ALIGN_DOWN(mrd->vbase);
+		len   = PAGE_ALIGN_UP(mrd->len + (mrd->vbase - vaddr));
 
-	/* Initialize the frame allocator by taking away the memory from the
-	 * larger heap area. We setup a new heap area later.
-	 */
-	if (_libkvmplat_cfg.heap2.len > _libkvmplat_cfg.heap.len) {
-		mr[0] = &_libkvmplat_cfg.heap2;
-		mr[1] = &_libkvmplat_cfg.heap;
-	} else {
-		mr[0] = &_libkvmplat_cfg.heap;
-		mr[1] = &_libkvmplat_cfg.heap2;
+		rc = ukplat_page_unmap(&kernel_pt, vaddr,
+				       len >> PAGE_SHIFT,
+				       PAGE_FLAG_KEEP_FRAMES);
+		if (unlikely(rc))
+			return rc;
 	}
 
-	offset = mr[0]->start - PAGE_ALIGN_UP(mr[0]->start);
-	start  = PAGE_ALIGN_UP(mr[0]->start);
-	len    = PAGE_ALIGN_DOWN(mr[0]->len - offset);
+	/* Perform mappings */
+	ukplat_memregion_foreach(&mrd, 0, UKPLAT_MEMRF_MAP,
+				 UKPLAT_MEMRF_MAP) {
+		UK_ASSERT(mrd->vbase != __U64_MAX);
 
-	rc = ukplat_pt_init(&kernel_pt, start, len);
-	if (unlikely(rc))
-		goto EXIT_FATAL;
+		vaddr = PAGE_ALIGN_DOWN(mrd->vbase);
+		paddr = PAGE_ALIGN_DOWN(mrd->pbase);
+		len   = PAGE_ALIGN_UP(mrd->len + (mrd->vbase - vaddr));
+		prot  = bootinfo_to_page_attr(mrd->flags);
 
-	/* Also add memory of the smaller heap region. Since the region might
-	 * be as small as a single page or less, we do not treat errors as
-	 * fatal here
-	 */
-	offset = mr[1]->start - PAGE_ALIGN_UP(mr[1]->start);
-	start  = PAGE_ALIGN_UP(mr[1]->start);
-	len    = PAGE_ALIGN_DOWN(mr[1]->len - offset);
+		rc = ukplat_page_map(&kernel_pt, vaddr, paddr,
+				     len >> PAGE_SHIFT, prot, 0);
+		if (unlikely(rc))
+			return rc;
+	}
 
-	ukplat_pt_add_mem(&kernel_pt, start, len);
+	return 0;
+}
 
-	/* Switch to new page table */
-	rc = ukplat_pt_set_active(&kernel_pt);
-	if (unlikely(rc))
-		goto EXIT_FATAL;
+static int mem_init(struct ukplat_bootinfo *bi)
+{
+	int rc;
 
-	/* Unmap all 1:1 mappings extending over the first megabyte, the kernel
-	 * image and initrd. The boot page table maps the first 4 GiB with
-	 * everything starting from 2 MiB mapped as 2 MiB large pages (see
-	 * pagetable64.S).
+	/* TODO: Until we generate the boot page table at compile time, we
+	 * manually add an untyped unmap region to the boot info to force an
+	 * unmapping of the 1:1 mapping after the kernel image before mapping
+	 * only the necessary parts.
 	 */
-	offset = mr[0]->start - PAGE_LARGE_ALIGN_UP(mr[0]->start);
-	start  = PAGE_LARGE_ALIGN_UP(mr[0]->start);
-	len    = PAGE_LARGE_ALIGN_DOWN(mr[0]->len - offset);
-
-	rc = ukplat_page_unmap(&kernel_pt, start,
-			       (PLATFORM_MAX_MEM_ADDR - start) >> PAGE_SHIFT,
-			       PAGE_FLAG_KEEP_FRAMES);
-	if (unlikely(rc))
-		goto EXIT_FATAL;
+	rc = ukplat_memregion_list_insert(&bi->mrds,
```

Comment:
With the plat-rearch we will move the paging API into a separate library. For that reason, I also designed the `paging_init()` to be self-contained, so it does not need additional data from the platform. Now, since I cannot do the unmap before (not initialized) or after (clashes with existing mappings) `paging_init()`, and I don't want to put the unmap in there explicitly (that would introduce platform knowledge about the 1:1 boot page table), the way to go is to use an UNMAP entry.

Actually, it is not really a dummy entry in the sense that it would be a hack here. In fact, I expect for chainloading we will have unmapping entries in there to unmap the first-stage loader after jumping into the (relocated) unikernel. What I might add is to remove unmap entries from the list after having done the unmap. But since they are usually not listed anyways, I did not include this for now.